### PR TITLE
Reduce allocations in RuntimeDescription and RuntimeDependencySet

### LINKED ISSUE #12714: Heavy allocations in NuGet.Commands.RestoreRunner.ExecuteAndCommitAsync|nuget.packaging.dll!NuGet.RuntimeModel.RuntimeDescription
https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1837525

### PR DESCRIPTION
**Pull request by drewnoakes**
## Bug

Fixes: https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1837525
Fixes: https://github.com/NuGet/Home/issues/12714

Regression? Last working version:

## Description

The `RuntimeDescription` and `RuntimeDependencySet` classes follow similar construction and clone patterns, and the changes to each class here are similar.

Previously each expected to receive non-null enumerables with input values, which would then be enumerated to materialise copies.

In cases where these collections were empty, the allocations are avoidable.

This changes the constructors to receive nullable values, allowing singleton collections to be used in such cases.

Similarly, when cloning, if the collections are empty, nulls are passed so that these singletons are applied. Also when cloning, the known collection type is cast to to enable allocation-free enumeration of the source when constructing the clone object.

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [ ] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [x] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

### CODE REVIEW

**Review by nkolev92 — DISMISSED** [review]

**Comment by nkolev92** [comment]
After a few of the other PRs were merged, there's conflicts @drewnoakes

**Comment by drewnoakes** [comment]
Merged in `dev` and resolved a few basic conflicts (we no longer have to clone runtimes due to #5249).

**Review by zivkan — APPROVED** [review]

File: src/NuGet.Core/NuGet.Packaging/RuntimeModel/RuntimeDescription.cs [hunk 1]
```diff
@@ -106,7 +116,11 @@ public static RuntimeDescription Merge(RuntimeDescription left, RuntimeDescripti
                 newSets[dependencySet.Id] = dependencySet;
             }
 
-            return new RuntimeDescription(left.RuntimeIdentifier, inheritedRuntimes, newSets.Values);
+            return new RuntimeDescription(
+                left.RuntimeIdentifier,
+                // If collections are empty, pass null to avoid allocations.
+                inheritedRuntimes.Count == 0 ? null : inheritedRuntimes,
+                newSets.Count == 0 ? null : newSets);
```
Comment (sbomer) [inline src/NuGet.Core/NuGet.Packaging/RuntimeModel/RuntimeDescription.cs:123]: `newSets` passed to the new overload will no longer be case-insensitive.

Comment (drewnoakes) [reply src/NuGet.Core/NuGet.Packaging/RuntimeModel/RuntimeDescription.cs:123]: Indeed! Fixed in https://github.com/NuGet/NuGet.Client/pull/5321

**Pull request merged**